### PR TITLE
[staking] fix ToEthTx() panic

### LINKED ISSUE #3909: candidateUpdate with nil operator causing web3 server to panic
### What version of iotex-core image (or code branch) are you using?
v1.11.0/master

### What operating system and processor architecture are you using?
mac

### What did you do? If possible, provide a recipe for reproducing the error.
```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","id":1,"method":"eth_getBlockByNumber", "params":["0xB9E688",true]}' https://babel-api.mainnet.iotex.io
```

### What did you expect to see?
web3 server should handle it normally

### What did you see instead?
```
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx</center>
<script>(function(){var js = "window['__CF$cv$params']={r:'7edf152d0d711726'};_cpo=document.createElement('script');_cpo.nonce='',_cpo.src='/cdn-cgi/challenge-platform/scripts/invisible.js',document.getElementsByTagName('head')[0].appendChild(_cpo);";var _0xh = document.createElement('iframe');_0xh.height = 1;_0xh.width = 1;_0xh.style.position = 'absolute';_0xh.style.top = 0;_0xh.style.left = 0;_0xh.style.border = 'none';_0xh.style.visibility = 'hidden';document.body.appendChild(_0xh);function handler() {var _0xi = _0xh.contentDocument || _0xh.contentWindow.document;if (_0xi) {var _0xj = _0xi.createElement('script');_0xj.innerHTML = js;_0xi.getElementsByTagName('head')[0].appendChild(_0xj);}}if (document.readyState !== 'loading') {handler();} else if (window.addEventListener) {document.addEventListener('DOMContentLoaded', handler);} else {var prev = document.onreadystatechange || function () {};document.onreadystatechange = function (e) {prev(e);if (document.readyState !== 'loading') {document.onreadystatechange = prev;handler();}};}})();</script></body>
```

### PR DESCRIPTION
**Pull request by dustinxie**
# Description
as title, issue found and reported by Jeremi

Fixes #3909 

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: action/candidate_register.go [hunk 1]
```diff
@@ -374,15 +380,11 @@ func ethAddrToNativeAddr(in interface{}) (address.Address, error) {
...
-	addr, err := address.FromString(address.StakingProtocolAddr)
-	if err != nil {
-		return nil, err
-	}
-	ethAddr := common.BytesToAddress(addr.Bytes())
 	data, err := cr.encodeABIBinary()
 	if err != nil {
 		return nil, err
 	}
+	ethAddr := common.BytesToAddress(address.StakingProtocolAddrHash[:])
```
Comment (dustinxie) [inline action/candidate_register.go:387]: easier way of getting the address, compared to 5 lines in the left side

File: action/candidate_update.go [hunk 2]
```diff
@@ -200,9 +200,15 @@ func (cu *CandidateUpdate) EncodeABIBinary() ([]byte, error) {
...
 func (cu *CandidateUpdate) encodeABIBinary() ([]byte, error) {
-	operatorEthAddr := common.BytesToAddress(cu.operatorAddress.Bytes())
-	rewardEthAddr := common.BytesToAddress(cu.rewardAddress.Bytes())
-	data, err := _candidateUpdateMethod.Inputs.Pack(cu.name, operatorEthAddr, rewardEthAddr)
+	if cu.operatorAddress == nil {
+		return nil, ErrAddress
+	}
+	if cu.rewardAddress == nil {
+		return nil, ErrAddress
+	}
```
Comment (dustinxie) [inline action/candidate_update.go:208]: missing these 2 checks caused panic

**Comment by millken** [comment]
good fix, I found in the ops exceptions log that `eth_getBlockByNumber` API generated panic log (runtime error: invalid memory address or nil pointer dereference), but I tested it N times, and I don't know what parameter triggered the bug.

File: action/stake_reclaim.go [hunk 3]
```diff
@@ -213,15 +213,11 @@ func NewUnstakeFromABIBinary(data []byte) (*Unstake, error) {
...
-	addr, err := address.FromString(address.StakingProtocolAddr)
-	if err != nil {
-		return nil, err
-	}
-	ethAddr := common.BytesToAddress(addr.Bytes())
 	data, err := su.encodeABIBinary()
 	if err != nil {
 		return nil, err
 	}
+	ethAddr := common.BytesToAddress(address.StakingProtocolAddrHash[:])
```
Comment (millken) [inline action/stake_reclaim.go:220]: It is used in several places, can place this in global var()？

File: action/builder.go [hunk 4]
```diff
@@ -22,7 +22,6 @@ type Builder struct {
 }
 
 var (
-	_stakingProtocolAddr, _   = address.FromString(address.StakingProtocolAddr)
```
Comment (Liuhaai) [inline action/builder.go:25]: filed in a new pr?

File: action/candidate_register.go (hunk 1 above)
Comment (Liuhaai) [reply action/candidate_register.go:387]: also in a new pr

**Comment by Liuhaai** [comment]
Bc no vriadic param is available in solidity, the query of native action is to be disabled without choices.

**Comment by Liuhaai** [comment]
Bc no vriadic param is available in solidity, the query of native action is to be disabled without choices.

File: action/candidate_update.go [hunk 5]
```diff
@@ -200,9 +200,15 @@ func (cu *CandidateUpdate) EncodeABIBinary() ([]byte, error) {
 }
 
 func (cu *CandidateUpdate) encodeABIBinary() ([]byte, error) {
-	operatorEthAddr := common.BytesToAddress(cu.operatorAddress.Bytes())
-	rewardEthAddr := common.BytesToAddress(cu.rewardAddress.Bytes())
-	data, err := _candidateUpdateMethod.Inputs.Pack(cu.name, operatorEthAddr, rewardEthAddr)
+	if cu.operatorAddress == nil {
+		return nil, ErrAddress
```
Comment (Liuhaai) [inline action/candidate_update.go:204]: it's not caused by an invalid address, but an option param

File: action/candidate_update.go (hunk 2 above)
Comment (dustinxie) [reply action/candidate_update.go:208]: yes, that is the native side, it has the null-check

this code is used to decode the txs in the block when returning a block via web3 format, it is missing the null-check

File: action/candidate_register.go [hunk 6]
```diff
@@ -304,14 +304,20 @@ func (cr *CandidateRegister) EncodeABIBinary() ([]byte, error) {
...
-	ownerEthAddr := common.BytesToAddress(cr.ownerAddress.Bytes())
+	if cr.operatorAddress == nil {
+		return nil, ErrAddress
+	}
+	if cr.rewardAddress == nil {
+		return nil, ErrAddress
+	}
+	if cr.ownerAddress == nil {
+		return nil, ErrAddress
+	}
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply action/candidate_register.go:315]: see above comment

File: action/candidate_register.go (hunk 1 above)
Comment (dustinxie) [reply action/candidate_register.go]: ok, will file a new PR

**Review by millken — APPROVED** [review]

File: action/candidate_update.go (hunk 2 above)
Comment (envestcc) [reply action/candidate_update.go:208]: ok, then should we return normally using ZeroAddress to represent the nil param, or just return error to damage and compatibility with eth?

**Pull request merged**